reference/src/layout/structs-and-tuples.md

```diff
@@ -122,6 +122,26 @@ compiler will not reorder it, to allow for the possibility of
 unsizing. E.g., `struct Foo { x: u16, y: u32 }` and `struct Foo<T> {
 x: u16, y: T }` where `T = u32` are not guaranteed to be identical.
 
+#### Zero-sized structs
+
+Structs with default layout are zero-sized, if they contain no fields of
+non-zero size. That is, either the type has no fields, or all of its fields have
+zero size.
+
+For example, all these types are zero-sized:
+
+```rust
+# use std::mem::size_of;
+struct Zst0;
+struct Zst1(Zst0);
+struct Zst2(Zst1, Zst0);
+# fn main() {
+# assert_eq!(size_of::<Zst0>(), 0);
+# assert_eq!(size_of::<Zst1>(), 0);
+# assert_eq!(size_of::<Zst2>(), 0);
+# }
+```
```

Comment:
I am feeling like we are increasingly listing weird special cases here and will soon be lost in the woods of them, where really these all fall out from the same pattern. :/

This overlaps with https://github.com/rust-lang/unsafe-code-guidelines/pull/164. I think we should try to find a more general statement that subsumes both.

OTOH, this here also applies if the ZST have nontrivial alignment, right? If yes, there should be an example of that.

Reply:
> where really these all fall out from the same pattern. :/

I'm open to suggestions. @rkruppe had the idea that we should document, e.g., all zero-sized related guarantees in some "central" place, so that one doesn't need to go through all layout documents looking for them. I agreed that this might make sense, once we have collected them all. At that point, the pattern might be obvious.

Reply:
> OTOH, this here also applies if the ZST have nontrivial alignment, right? If yes, there should be an example of that.

Yes, this applies for all alignments. I"ll modify the example to show that.